raft: fix leadership instability on follower node restarts

jcsp opened the pull request:
## Cover letter

Our raft implementation is supposed to keep leadership stable through follower node restarts, but in certain situations the leader was non-deterministically changing after a follower restart.  This was visible to users as a confusing behaviour where sometimes after a node restart, the restarted node would regain leadership of partitions (unintended), whereas sometimes the restarted node would not be leader for anything after restart (intended).

This issue was found in new ducktape tests for raft.  These are still being added to, but it makes sense to include the current state in this PR as a regression test for the leader stability fix.

Fixes: https://github.com/vectorizedio/redpanda/issues/2048

## Release notes

In some situations, leadership of a topic's raft group could change unexpectedly when a non-leader node restarts.  After this change, leadership should remain stable through follower node restarts, notwithstanding any intentional leadership changes orchestrated by the new leadership balancing feature.

Lazin commented on src/v/raft/consensus.cc:
```diff
@@ -1227,6 +1227,29 @@ ss::future<vote_reply> consensus::do_vote(vote_request&& r) {
+                  "Vote request from peer {} with heartbeat failures, "
+                  "resetting backoff",
+                  r.node_id);
+                return _client_protocol.reset_backoff(r.node_id.id())
```
What if the node is not restarted but just unresponsive?

jcsp replied:
I think the same logic should apply in both cases: the vote/prevote message is a signal that the node is alive, and therefore we shouldn't be backing off from sending it heartbeats.

Lazin reviewed (approved):
LGTM

dotnwat commented on src/v/raft/consensus.cc:
```diff
@@ -1335,6 +1328,13 @@ ss::future<vote_reply> consensus::do_vote(vote_request&& r) {
+    reply.log_ok
+      = r.prev_log_term > last_entry_term
+        || (r.prev_log_term == last_entry_term && r.prev_log_index >= last_log_index);
+
```
ahhh, I think this change partially addresses the question on the previous commit.

dotnwat reviewed (commented):
These changes look good to me. Sorry about the delay reviewing! @mmaslankaprv @rystsov please take a look too.

jcsp commented on src/v/raft/consensus.cc:
```diff
@@ -1266,6 +1266,29 @@ ss::future<vote_reply> consensus::do_vote(vote_request&& r) {
+    // to have been recently restarted (have failed heartbeats
+    // and an <= present term), reset their RPC backoff to get
+    // a heartbeat sent out sooner.
+    if (is_leader() and r.term <= _term) {
```
As you notice in the later commit, the case where we were potentially telling the candidate log_ok=true even for old terms was a bug, letting prevoters advance to vote when they shouldn't.

 In general the condition for comparing terms for votes is `<` but when we're already a leader in the current term, I don't think it's ever valid to approve someone else voting for themselves in the same term.  But we do still want to reset the follower's backoff if they're trying to vote in a term where we're already the leader.

rystsov replied:
It's more a feature than a bug - https://github.com/vectorizedio/redpanda/pull/2071/commits/380af4a73a8e02d7699c68ee710906fdf0f619cb#r705736869

rystsov commented on src/v/rpc/connection_cache.h:
```diff
@@ -116,6 +116,25 @@ class connection_cache final
+                  return;
+              }
+              auto recon_transport = cache.get(node_id);
+              recon_transport->reset_backoff();
```
It seems that `reset_backoff` isn't defined yet, you add it only in c336f5f. Can you please make each commit compile-able so we could use `git bisec` if we need to in the future.

rystsov commented on src/v/raft/consensus.cc:
```diff
@@ -1289,13 +1289,6 @@ ss::future<vote_reply> consensus::do_vote(vote_request&& r) {
-    /// set to true if the caller's log is as up to date as the recipient's
-    /// - extension on raft. see Diego's phd dissertation, section 9.6
-    /// - "Preventing disruptions when a server rejoins the cluster"
-    reply.log_ok
```
This looks very like a recipe for a deadlock. As I remember we **must** set log_ok even when `r.term < _term`.

There is `prevote_stm` optimization. We use it to check that the  current candidate's log is up to date with the majority of the nodes before starting an election round (when the candidate isn't up to date it it doesn't make sense to inclement its term and to disrupt the current leader).

Without log_ok we can't increment term, without incrementing term we can't pass `r.term < _term`, without passing it we can't set log_ok => deadlock

mmaslankaprv replied:
This is valid comment. Would it be possible to rework prevote request handing in a way that it will return `granted` flag based on the time since last heartbeat ? It will be independent from the term and will not cause the deadlock

jcsp commented:
I've changed how this works to keep the desired behaviour of updating log_ok for voters with old terms so that they can increment their term and try again.

Now we just rely on the existing "Stable leadership optimization" code block, but move it ahead of the update to log_ok so that nodes with recent heartbeats from the leader will not approve prevotes (where approval is implicit in updating log_ok).

Reasoning for the top two checks in the function as follows:

Rejecting votes with `if (is_leader() and r.term <= _term) ` && failed heartbeats to the voter
- This only makes the leader reject the vote: another node can still win the vote with a quorum, and displace this leader with an append_entries message with a higher term.  If the voter and the leader are the only two nodes, then the voter can never win the election, it must wait for an append entries from the leader.  If there are any other nodes, then they will decide whether to approve the prevotes based on the next `_hbeat > prev_election` condition, and the leader may be displaced if for some reason its heartbeats weren't reaching the other nodes.

Rerjecting votes if `_hbeat > prev_election`:
- This is part of the original prevote definition in (Diego Ongaro's thesis): prevotes should be approved "if the candidate’s log is sufficiently up-to-date, and the voters have not received heartbeats from a valid leader for at least a baseline election timeout".  Previously we were only using the first part of the definition, because we updated the log_ok field before checking if the node had received a recent heartbeat.

jcsp commented:
The test failures are the node_operations_fuzzy_test ones that were fixed on dev in #2338

dotnwat commented:
@mmaslankaprv @rystsov i think this is ready for another review

mmaslankaprv reviewed (approved):
LGTM

rystsov reviewed (approved):
LGTM
